third_party/validator/cai.go

```diff
@@ -34,3 +41,70 @@ type IAMBinding struct {
 	Role    string   `json:"role"`
 	Members []string `json:"members"`
 }
+
+// assetName templates an asset.name by looking up and replacing all instances
+// of {{field}}. In the case where a field would resolve to an empty string, a
+// generated unique string will be used: "<placeholder-" + randomString() + ">".
+// This is done to preserve uniqueness of asset.name for a given asset.asset_type.
+func assetName(d TerraformResourceData, config *Config, linkTmpl string) (string, error) {
```

Comment:
@nstogner This looks like a lot of duplicated logic from [replaceVars](https://github.com/GoogleCloudPlatform/magic-modules/blob/cd00e01eafc515d5d1029b5eaeb064a094dd4850/third_party/terraform/utils/transport.go#L122). Could we somehow try to keep them consolidated?

Reply:
Good point, I don't know what I thinking there, fixing now

Reply:
Refactored `replaceVars`